---
title: "01_collate_from_all_server_shotgun"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(readxl)
library(kableExtra)
```

## Gather project ID number from different spreadsheets

```{r}
#  castoricenter server 
castori_proj <- read_excel('/Volumes/castoricenter/Human.Sequencing.Data/Human.Shotgun.Sequenced.xlsx') %>% 
  # there are pending samples that are currently being sequenced
  filter(`Shotgun.Sequenced` == 'yes') %>% 
  distinct(`iGO.Project.Number`) %>% 
  pull(`iGO.Project.Number`)
```


```{r}
# Gabe  
vdb_proj <- read_excel('../data/VDB.Human.Shotgun.Sequenced.xlsx') %>% 
  transmute(iGO_Project_Number = `iGO.Project.Number`)  %>% 
  distinct(iGO_Project_Number) %>% 
  mutate(iGO_Project_Number = str_glue('Project_{iGO_Project_Number}')) %>% 
  pull(iGO_Project_Number)

# Annie
annie_proj <- paste0('Project_', c(
  '08788',
  '08788_B',
  '08788_C',
  '09050',
  '09102',
  '07051_C',
  '08290',
  '09099_B'), sep = '' )

# Melody and Tsoni's server . probably some of them have been copied to vdb server. will remove duplicates later

other_proj <- c('/Volumes/smithm4/FASTQ/','/Volumes/peledj/FASTQ/', '/Volumes/ponced/FASTQ/') %>% 
  map_dfr(~ Sys.glob(str_glue('{.}*')) %>% 
            data_frame() %>% 
            rename( path = names(.)[1])) %>% 
  transmute(iGO_Project_Number = str_extract(path, 'Project_.+$')) %>% 
  pull(iGO_Project_Number)



# all the unique project ID numbers 
all_proj <- unique(c(castori_proj, vdb_proj, annie_proj, other_proj, cart))
  
```



## Retrieve the file path of the samples from servers

Make sure fully connected to all servers

```{r}
# list all servers
servers <- tibble(server = Sys.glob('/Volumes/*') ) %>% 
  filter(!str_detect(server, 'Macintosh|castor')) %>% 
  mutate(folder = if_else(str_detect(server, 'vandenBrinkLab'), str_glue('{server}/deep_sequencing'),str_glue('{server}/FASTQ')))

# find all of the projects in all of the servers

ALL <- servers %>% 
  pull(folder) %>% 
  map_dfr(function(f){
    all_proj %>% 
      set_names(all_proj) %>% 
      imap_dfr( ~ Sys.glob(str_glue('{f}/{..1}/*/*')) %>% 
             data_frame %>% 
              mutate(..2)) %>% 
      rename(directory = names(.)[1],
             projectID = names(.)[2])
  }) %>% 
  # trying to remove the duplicated copied samples here
  mutate(sample_folder = str_extract(directory, 'Sample_.+$'),
         fid = str_replace(sample_folder, 'Sample_',''),
         fid = str_replace(fid, '_IGO_.+$','')) %>% 
  distinct(sample_folder, .keep_all = T)

```

## Create the sampleid from the sample_folder and clean the FMT sampleid here

```{r}
splitFMT <- ALL %>% 
   mutate(sampleid = str_replace(sample_folder, 'Sample_',''),
          sampleid = str_replace(sampleid, '_IGO_.+$','')) %>% 
  split(f = str_detect(.$sampleid, '^FMT'))

# below the cleaned sampleid df
ALL_cleaned <- bind_rows(
  splitFMT[['TRUE']] %>% 
    separate(sampleid, into = c('other','part'), sep= 'FMT') %>% 
    # remove the _ cuz its never in the sampleid
    mutate(part = str_replace(part, '^_','')) %>% 
    # str_extract to get the numbers and letters individually
    mutate(num = str_extract(part, '[0-9]+'),
           lett = str_extract(part, '[aA-zZ]+'),
           num = str_remove(num, '^0+')) %>% 
    select(-other, -part) %>% 
    mutate(num = str_pad(num, 4, 'left', '0'),
           sampleid = str_glue('FMT.{num}{lett}')) %>% 
    select(directory, projectID, sampleid, fid),
  splitFMT[['FALSE']] %>% 
    select(directory, projectID, sampleid, fid)
  )


```

## Continue cleaning on some really messy weird sampleids format

```{r}
# the 34/105 samples are wrongly named
c34 <- ALL_cleaned %>% 
  split(f = str_detect(.$sampleid, '^34'))

ALL_cleaned <- bind_rows(
  c34[['FALSE']],
  c34[['TRUE']] %>% 
    mutate(sampleid = str_glue('FMT.00{sampleid}'))
)


# there is one sample 1058C it's a valid sample by itself, shouldn't add the below to it
# the below should only work for the samples in Project_07976
split_proj <- ALL_cleaned %>% 
  split(f = str_detect(.$projectID, 'Project_07976'))
 

ALL_cleaned <-  bind_rows(
  split_proj %>% 
    pluck('FALSE'),
  bind_rows(split_proj %>% 
    pluck('TRUE') %>% 
    split(f = str_detect(.$sampleid, '^105')) %>% 
    pluck('TRUE') %>% 
    mutate(sampleid = str_glue('FMT.0{sampleid}')),
  split_proj %>% 
    pluck('TRUE') %>% 
    split(f = str_detect(.$sampleid, '^105')) %>% 
    pluck('FALSE'))
)

# remove the samples with the Y... FC... MM... SOHN
ALL_cleaned <- ALL_cleaned %>% 
  filter(!str_detect(sampleid, '^Y|^MM|^FC|SOHN'))


# for some weird samples in 07051_C remove the leading X and the part after the undersore
ALL_cleaned <- ALL_cleaned %>% 
  mutate(sampleid = str_replace(sampleid, '^X', ''),
         sampleid = str_replace(sampleid, '-.+$', ''),
         sampleid = str_replace(sampleid, '_.+$', '')) 

# remove something in the peledj smithm4 server 
ALL_cleaned <- ALL_cleaned %>% 
  filter(!projectID %in% c('Project_09100','Project_09684_B','Project_10195', 'Project_08868', 
                           'Project_08899', 'Project_08867', 'Project_08952', 'Project_08989', 'Project_09023'))


# theres some more cleaning to do, for example some samples in project 08290 should have FMT but missed that
split_8290 <- ALL_cleaned %>% 
  split(f = str_detect(.$projectID, 'Project_08290'))

ALL_cleaned <-  bind_rows(
  split_8290 %>% 
    pluck('FALSE'),
  bind_rows(split_8290 %>% 
    pluck('TRUE') %>% 
    split(f = str_detect(.$sampleid, '^0')) %>% 
    pluck('TRUE') %>% 
    mutate(sampleid = str_glue('FMT.{sampleid}')),
  split_8290 %>% 
    pluck('TRUE') %>% 
    split(f = str_detect(.$sampleid, '^0')) %>% 
    pluck('FALSE'))
)


# there is really this weird sample 1058C not sure why my algorithm would add "FMT" in front of it

```

```{r}
# write out the final cleaned table

ALL_cleaned %>% 
  write_csv('/Volumes/vandenBrinkLab/Angel_Dai/Full_human_shotgun_catalog/full_human_shotgun_catalog_updated.csv')

```

```{r}
ALL_cleaned <- read_csv('/Volumes/vandenBrinkLab/Angel_Dai/Full_human_shotgun_catalog/full_human_shotgun_catalog_updated.csv') %>% 
  mutate(fid = str_replace(directory, '^.+Sample_',''),
         fid = str_replace(fid, '_IGO.+$',''))


ALL_cleaned %>% 
  head(10) %>% 
  kable(caption = 'A glimpse of the finalized catalog') %>% 
  kable_styling(full_width = F, position = 'left')
```
```{r}
read_csv('~/projects/Doris/analysis/data/shotgun_lookup.csv') %>% 
  head(10) %>% 
  kable(caption = 'Previous shotgun lookup table') %>% 
  kable_styling(full_width = F, position = 'left')
```

## the newly added since 2019 Nov

```{r}
# 2020-11-12
# the shotgun from italy and upenn in the CART project
cart <- c('Project_11341','Project_11038')
cart <- c('Project_11038')

cart2 <- tibble(
  folder = '/Volumes/vandenBrinkLab/deep_sequencing'
)

cart2_files <- cart2 %>% 
  pull(folder) %>% 
  map_dfr(function(f){
    cart %>% 
      set_names(cart) %>% 
      imap_dfr( ~ Sys.glob(str_glue('{f}/{..1}/*/*')) %>% 
             data_frame %>% 
              mutate(..2)) %>% 
      rename(directory = names(.)[1],
             projectID = names(.)[2])
  }) %>% 
  mutate(sample_folder = str_extract(directory, 'Sample_.+$'),
         fid = str_replace(sample_folder, 'Sample_',''),
         fid = str_replace(fid, '_IGO_.+$','')) 

# duplicated sample folders??
dups <- cart2_files %>% 
  filter(duplicated(sample_folder))

dups %>% 
  write_csv('../data/duplicated_samples_from_italy_cohort.csv')

cart2_files %>% 
  write_csv('../data/full_project_11308_files.csv')
```



```{r}
# checking whether a list of samples are already shotgun sequenced. 
source('~/db_connect_simple.R')
connect_database(config_file = '~/dbConfig.txt')
get_table_from_database('shotgun_lookup_ad')
samps <- read_table('~/Downloads/samples.txt', col_names = F)

shotgun_lookup_ad %>% 
  filter(sampleid %in% samps$X1) %>% 
  distinct(sampleid, .keep_all = T) %>% 
  write_csv('../data/Chen_request.csv')

```


```{r}
# the several samples in the IL22 of Doris project 
project <- 'Project_11280'
doris_new <-  '/Volumes/ponced/FASTQ/' %>% 
  map_dfr(function(f){
    project %>% 
      set_names(project) %>% 
      imap_dfr( ~ Sys.glob(str_glue('{f}/{..1}/*/*')) %>% 
             data_frame %>% 
              mutate(..2)) %>% 
      rename(directory = names(.)[1],
             projectID = names(.)[2])
  }) %>% 
  mutate(sample_folder = str_extract(directory, 'Sample_.+$'),
         fid = str_replace(sample_folder, 'Sample_',''),
         fid = str_replace(fid, '_IGO_.+$','')) %>% 
  mutate(sampleid = fid) %>% 
  select(-sample_folder) %>% 
  select(directory, projectID, sampleid, fid)

cart <- files %>% 
  select(directory, projectID, sampleid, fid)
  

new  <- bind_rows(
  cart,
  doris_new
)

new %>% 
  write_csv('/Volumes/vandenbrinklab/Angel_Dai/Full_human_shotgun_catalog/shotgun_update_1204.csv')
  

source('../../../microbiome_db/SQL/scripts/upload_data_from_query_OTU_AD.R')
upload_data_from_query_OTU(1) # changed the file path of the table getting uploaded

source('~/db_connect_simple.R')
connect_database(config_file = '~/dbConfig.txt')
get_table_from_database('shotgun_lookup_ad')
```


