---
title: "A function to get new shotgun sample info and then update the lookup table"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library(tidyverse)
```

A function to get new shotgun sample info and then update the lookup table

```{r}
# the current shotgun lookup table 
library(vdbR)
connect_database(config_file = '~/dbConfig.txt')
get_table_from_database('shotgun_lookup_ad')
get_table_from_database('samples_castori_ag')

```

**BE SURE TO CONNECT TO ALL SERVERS BEFORE LOOKING FOR ANYTHING!!!!!**

```{r}
# input: 
# dir could be "/Volumes/ponced/FASTQ/" or "/Volumes/vandenBrinkLab/deep_sequencing"
# projectid: "Project_11280"

# output: a table in the format of shotgun_lookup_ad with 4 columns: directory, projectID, sampleid, fid

# NEEDS TO MANUALLY CHECK THE SAPMLEID THAT IS DERIVED FROM THE FID, COULD IT POSSIBLY BE THE REAL SAMPLEID?
# THIS WORKS BEST WHEN PROCESSING SAMPELS FROM GABE SINCE EMILY HAS A WELL KEPT TABLE I CAN USE 

new_shotgun_sample_tb <- function(projectid_, dir_){
  
  ret =  dir_ %>% 
    map_dfr(function(f){
      projectid_ %>% 
        set_names(projectid_) %>% 
        imap_dfr( ~ Sys.glob(str_glue('{f}/{..1}/*/*')) %>% 
               tibble %>% 
                mutate(..2)) %>% 
        rename(directory = names(.)[1],
               projectID = names(.)[2])
    }) %>% 
    mutate(sample_folder = str_extract(directory, 'Sample_.+$'),
           fid = str_replace(sample_folder, 'Sample_',''),
           fid = str_replace(fid, '_IGO_.+$','')) %>% 
    mutate(sampleid = fid) %>% 
    select(-sample_folder) %>% 
    select(directory, projectID, sampleid, fid) %>% 
    distinct(sampleid, .keep_all = T)
}


```

```{r}
# 2020-12-7
# look at the castori table from Emily
length(intersect(castori_proj$Sample.ID, shotgun_lookup_ad$sampleid))
length(setdiff(castori_proj$Sample.ID, shotgun_lookup_ad$sampleid))
 
# to get the new samples that need to be uploaded and changed to the right format
new_castori_samples <- castori_proj %>% 
  distinct(Sample.ID, .keep_all = T) %>% 
  filter(! Sample.ID %in% shotgun_lookup_ad$sampleid) %>% 
  filter(!str_detect(Sample.ID, '^DP'))   # already included in the current lookup table without dot in the sampleid


new_castori_df <- new_castori_samples %>%   
  mutate(PI_name = str_replace(...6, ' server','')) %>% 
  mutate(server = str_glue('/Volumes/{PI_name}/FASTQ/')) %>% 
  distinct(server, iGO.Project.Number) %>% 
  rename(dir = server, projectid = iGO.Project.Number)

new <- purrr::pmap_dfr(new_castori_df, function(projectid, dir){
    new_shotgun_sample_tb(projectid, dir)
  })

new %>% 
  write_csv('/Volumes/vandenbrinklab/Angel_Dai/Full_human_shotgun_catalog/shotgun_update_1207.csv')


new_castori_samples %>% 
  dplyr::count(iGO.Project.Number) # one in Project_10799 didn't actually return the file

```

```{r}
# 2020-12-8 the new samples from italy and upenn
two <- new_shotgun_sample_tb('Project_11453', '/Volumes/smithm4/FASTQ/') 

two %>% 
  write_csv('/Volumes/vandenbrinklab/Angel_Dai/Full_human_shotgun_catalog/shotgun_update_1208.csv')

file.edit('../../../microbiome_db/SQL/scripts/upload_data_from_query_OTU_AD.R')
```

```{r}
# 2020-12-8 fixing the mislabelled samples in 10754 by uploading the vdb server version of it
fix <- new_shotgun_sample_tb('Project_10754', '/Volumes/vandenBrinkLab/deep_sequencing')

fix %>% 
  write_csv('/Volumes/vandenbrinklab/Angel_Dai/Full_human_shotgun_catalog/shotgun_update_1208.csv')
```

```{r}
# 2021-3-9 the 500 samples from Oriana batch
# trying to see if the fid is the sampleid 
list_table_from_database('castori')
get_table_from_database('samples_castori_ag')

O <- new_shotgun_sample_tb('Project_11592', '/Volumes/vandenbm/FASTQ/') 

O_list <- O %>% 
  split(.$sampleid %in% samples_castori_ag$sampleid)

fix <- O_list %>% 
  pluck('FALSE') %>% 
  mutate(sampleid = str_replace(sampleid, 'FMT','FMT.'),
         sampleid = str_replace(sampleid, '^0','')) 

# double check if they are now all the correct sampleid
fix %>% 
  filter(sampleid %in% samples_castori_ag$sampleid) %>% 
  nrow

O_fix <- bind_rows(
  O_list %>% pluck('TRUE'),
  fix
) %>% 
  filter(!sampleid %in% shotgun_lookup_ad$sampleid) %>% 
  # remove one duplicate that is 0116E
  arrange(sampleid, desc(fid)) %>% 
  distinct(sampleid, .keep_all = T)


O_fix %>% 
  write_csv('../data/Oriana_480_update.csv')

# write out a script that I can rsync to cluster 
shotgun_lookup_ad %>% 
  filter(projectid == 'Project_11592') %>% 
  mutate(cmd = str_glue('rsync --progress --partial -avz {directory} daia1@lilac.mskcc.org:~/my_workdir/samples')) %>% 
  select(cmd) %>% 
  write_csv('../data/O480_rsync.sh', col_names = F)

# only get the fid of these 480 samples 
samples <- read_csv('../data/Oriana_480_update.csv') %>% 
  select(fid) %>% 
  write_csv('../data/Oriana_480_fid.csv', col_names = F)
  
```

```{r}
# Arielle shotgun samples
a <- readxl::read_excel('../data/FMT_case_series_sample_IDs.xlsx') %>% 
  mutate(fid = str_replace_all(sampleid, '\\.','_')) %>% 
  mutate(fid = str_replace(fid, 'VDB','vdB'))

a %>% 
  write_csv('../data/FMT_case_series_sample_IDs_match_fid.csv')

a %>% 
  select(fid) %>% 
  write_csv('../data/Arielle_fid.csv', col_names = F)

arielle <- tibble(
  directory = Sys.glob('/Volumes/castoricenter/Human.Sequencing.Data/ShotgunSeq/Hiseq.36/*'),
  projectID = 'Project_11606'
) %>% 
  mutate(sample_folder = str_extract(directory, 'Sample_.+$'),
           fid = str_replace(sample_folder, 'Sample_',''),
           fid = str_replace(fid, '_IGO_.+$','')) %>% 
  mutate(sampleid = fid) %>% 
  select(-sample_folder) %>% 
  mutate(sampleid = str_replace_all(fid, '_','\\.')) %>% 
  mutate(sampleid = str_replace(sampleid, 'vdB.FMT','VDB.FMT')) %>% 
  select(directory, projectID, sampleid, fid)
length(intersect(arielle$sampleid, a$sampleid))


```


```{r}
# my new shotgun samples
me <- new_shotgun_sample_tb('Project_11862', '/Volumes/vandenbm/FASTQ/') 

# out put the fid to the cluster
me %>% 
  select(fid) %>% 
  write_csv('../data/me_nutrition.csv', col_names = F)


# clean the sampleid so that it could be added to the shotgun lookup table
me_u <- me %>% 
  mutate(sampleid = str_replace(sampleid, 'FMT_','FMT.'))
length(intersect(me_u$sampleid, samples_castori_ag$sampleid))
```

```{r}
# 2021-4-20 Sarah's mouse data
sarah <- new_shotgun_sample_tb('Project_11861', '/Volumes/vandenbm/FASTQ/') 
```
```{r}
# all together 2021-4-20 to be updated 
today <- bind_rows(
  sarah, arielle, me_u
)
today %>% 
  write_csv('../data/new_20210420.csv')
```


```{r}
# update it to the db
#sometimes it doesn't work you just need to restart R and run it again 

# change the d_set_input in the upload_data_from_query_OTU_AD.R script to the new file that is saved above
# run the script upload_data_from_query_OTU_AD.R
upload_data_from_query_OTU(1) # changed the file path of the table getting uploaded 
```


